tests/integration/README.md

```diff
@@ -31,44 +42,338 @@ to write environment-agnostics tests in a high-level fashion. The quickest way t
 [examples](https://github.com/istio/istio/tree/master/tests/integration/examples) folder.
 
-The test framework has various flags that can be set to change its behavior.
+## Writing Tests
 
- 
-### Environments
+This section describes how to add your own suite and tests. You can find a more comprehensive examples in
+the [examples](https://github.com/istio/istio/tree/master/tests/integration/examples) folder.
+
+### Adding a Test Suite
+
+All tests that use the framework, must run as part of a *suite*. Only a single suite can be defined per package, since
+it is bootstrapped by a Go `TestMain`, which has the same restriction.
+
+To begin, create a new folder for your suite under 
+[tests/integration](https://github.com/istio/istio/tree/master/tests/integration).
+
+```console 
+$ cd ${ISTIO}/tests/integration
+$ mkdir mysuite
+```
+Within that package, create a `TestMain` to bootstrap the test suite:
+
+```go
+func TestMain(m *testing.M) {
+	framework.
+		NewSuite("mysuite", m).
+		Run()
+}
+```
+
+Next, define your tests in the same package:
 
-The test framework currently supports two environments:
+```go
+func TestMyLogic(t *testing.T) {
+	framework.
+		NewTest(t).
+		Run(func(ctx framework.TestContext) {
+			// Create a component
+			g := galley.NewOrFail(ctx, ctx, cfg)
+			
+			// Use the component.
+			g.ApplyConfigOrFail(ctx, nil, mycfg)
+			defer g.DeleteConfigOrFail(ctx, nil, mycfg)
+			
+			// Do more stuff here.
+		}
+}
+```
+
+The `framework.TestContext` is a wrapper around the underlying `testing.T` and implements the same interface. Test code
+should generally not interact with the `testing.T` directly.
+
+
+In the `TestMain`, you can also restrict the test to particular environment, apply labels, or do test-wide setup, such as
+ deploying Istio.
  
-  * **Local**: The test binaries run either in-memory, or locally as processes.
-  * **Kubernetes**: The test binaries run in a Kubernetes cluster, but the test logic runs in the test binary.
-  
-When running tests, only one environment can be specified:
+```go
+func TestMain(m *testing.M) {
+	framework.
+		NewSuite("mysuite", m).
+		// Restrict the test to the K8s environment only, tests will be skipped in native environment.
+		RequireEnvironment(environment.Kube).
+		// Deploy Istio on the cluster
+		Setup(istio.SetupOnKube(nil, nil)).
+		// Run your own custom setup
+		Setup(mySetup).
+		Run()
+}
 
-```console
-$ go test ./... -istio.test.env native
-$ go test ./... -p 1 -istio.test.env kube
+func mySetup(ctx framework.SuiteContext) error {
+    // Your own setup code	
+}
+ ```
+
+### Sub-Tests
+
+Go allows you to run sub-tests with `t.Run()`. Similarly, this framework supports nesting tests with `ctx.NewSubTest()`:
+
+```go
+func TestMyLogic(t *testing.T) {
+	framework.
+		NewTest(t).
+		Run(func(ctx framework.TestContext) {
+		
+			// Create a component
+            g := galley.NewOrFail(ctx, ctx, cfg)
+            			
+			configs := []struct{
```

Comment:
Nit: Unnecessary indentation.

Reply:
fixed ... it was actually due to a mix of tabs and spaces. Switched everything over to spaces.